src/Coordination/Changelog.cpp

```diff
@@ -68,80 +75,344 @@ Checksum computeRecordChecksum(const ChangelogRecord & record)
-        writeIntBinary(record.header.term, getBuffer());
-        writeIntBinary(record.header.value_type, getBuffer());
-        writeIntBinary(record.header.blob_size, getBuffer());
+            file_buf = std::make_unique<WriteBufferFromFile>(
+                file_description->path, DBMS_DEFAULT_BUFFER_SIZE, mode == WriteMode::Rewrite ? -1 : (O_APPEND | O_CREAT | O_WRONLY));
+            last_index_written.reset();
+            current_file_description = std::move(file_description);
 
-        if (record.header.blob_size != 0)
-            getBuffer().write(reinterpret_cast<char *>(record.blob->data_begin()), record.blob->size());
+            // try fixing ZstdStream for the file if we are appending to some old file
+            if (compress_logs && mode == WriteMode::Append && ZstdDeflatingAppendableWriteBuffer::isNeedToAddEmptyBlock(current_file_description->path))
+            {
+                ZstdDeflatingAppendableWriteBuffer::addEmptyBlock(*file_buf);
+                flush();
+            }
+
+            prealloc_done = false;
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log);
+            throw;
+        }
     }
 
-    void flush(bool force_fsync)
+    bool isFileSet() const { return file_buf != nullptr; }
+
+
+    bool appendRecord(ChangelogRecord && record)
     {
-        if (compressed_buffer)
+        const bool log_is_complete = record.header.index - getStartIndex() == current_file_description->expectedEntriesCountInLog();
+
+        if (log_is_complete)
+            rotate(record.header.index);
+
+        if (!prealloc_done) [[unlikely]]
         {
-            /// Flush compressed data to WriteBufferFromFile working_buffer
-            compressed_buffer->next();
+            tryPreallocateForFile();
+
+            if (!prealloc_done)
+                return false;
         }
 
-        WriteBuffer * working_buf = compressed_buffer ? compressed_buffer->getNestedBuffer() : file_buf.get();
+        const auto write_record = [&]
+        {
+            writeIntBinary(computeRecordChecksum(record), getBuffer());
+
+            writeIntBinary(record.header.version, getBuffer());
+
+            writeIntBinary(record.header.index, getBuffer());
+            writeIntBinary(record.header.term, getBuffer());
+            writeIntBinary(record.header.value_type, getBuffer());
+            writeIntBinary(record.header.blob_size, getBuffer());
 
-        /// Flush working buffer to file system
-        working_buf->next();
+            if (record.header.blob_size != 0)
+                getBuffer().write(reinterpret_cast<char *>(record.blob->data_begin()), record.blob->size());
 
+            if (compressed_buffer)
+            {
+                /// Flush compressed data to WriteBufferMemory working_buffer
+                compressed_buffer->next();
+            }
+        };
+
+        // write record to an in-memory block
+        write_record();
```

Comment:
We could simply reserve `2 * max_log_file_size` (and assume that one record cannot exceed `max_log_file_size`) to avoid creating that temporary buffer, but okay

Reply:
It will help only if one record cannot exceed `max_log_file_size` (probably it can in extremely rare cases, but I guess it could be ignored):
 - reserve `2 * max_log_file_size` when creating a log file
 - write directly to the file
 - check the actual size of the file after appending each record (or count how many bytes were written through `WriteBufferFromFile` )
 - rotate the log if its size exceeds `max_log_file_size`

So the file size will be less than `2 * max_log_file_size` if the assumption is correct.

Reply:
I see, that does make sense.
I would say an okay log size limit would be few 10s of MB (we can write a warning if the size limit is too small) so I can't imagine a record of that size.

I'll try to write something using that method (should allow me to avoid extra logic I added in other buffers 🥲 ) and see if there are any problems with that approach before addressing other comments.

Reply:
What do you think to have a fixed size we append to the `max_log_size` instead of doubling the size?
Let's say around 50MBs. That way we don't overallocate for larger `max_log_size` and don't allocate too little for small `max_log_size`

Reply:
Yes, it's a good idea